apps/api/v2/src/modules/oauth/oauth-client.repository.ts

```diff
@@ -54,4 +58,143 @@ export class OAuthClientRepository {
+        secret: input.client_secret,
+        authorizationTokens: {
+          some: {
+            id: tokenId,
+          },
+        },
+      },
+    });
+
+    if (!oauthClient) {
+      throw new BadRequestException("Invalid Authorization Token.");
+    }
+
+    const authorizationToken = await this.dbRead.prisma.platformAuthorizationToken.findUnique({
+      where: {
+        id: tokenId,
+      },
+      include: {
+        owner: {
+          select: {
+            id: true,
+          },
+        },
+      },
+    });
+
+    const accessExpiry = DateTime.now().plus({ days: 1 }).startOf("day").toJSDate();
+    const refreshExpiry = DateTime.now().plus({ year: 1 }).startOf("day").toJSDate();
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const [_, accessToken, refreshToken] = await this.dbWrite.prisma.$transaction([
+      this.dbWrite.prisma.platformAuthorizationToken.delete({ where: { id: tokenId } }),
+      this.dbWrite.prisma.accessToken.create({
+        data: {
+          secret: this.jwtService.sign(JSON.stringify({ type: "access_token", clientId: oauthClient.id })),
+          expiresAt: accessExpiry,
+          client: { connect: { id: input.client_id } },
+          owner: { connect: { id: authorizationToken?.owner.id } },
+        },
+      }),
+      this.dbWrite.prisma.refreshToken.create({
+        data: {
+          secret: this.jwtService.sign(JSON.stringify({ type: "refresh_token", clientId: oauthClient.id })),
+          expiresAt: refreshExpiry,
+          client: { connect: { id: input.client_id } },
+          owner: { connect: { id: authorizationToken?.owner.id } },
+        },
+      }),
+    ]);
+
+    void this.oauthService.propagateAccessToken(accessToken); // voided as we don't need to await
+
+    return {
+      access_token: accessToken.secret,
+      refresh_token: refreshToken.secret,
+    };
+  }
+
+  async refreshToken(clientId: string, clientSecret: string, tokenSecret: string) {
+    const oauthClient = await this.dbRead.prisma.platformOAuthClient.findFirst({
+      where: {
+        id: clientId,
+        secret: clientSecret,
+      },
+    });
+
+    if (!oauthClient) {
+      throw new BadRequestException("Invalid OAuthClient credentials.");
+    }
+
+    const expiring = await this.dbRead.prisma.refreshToken.findUnique({
+      where: {
+        secret: tokenSecret,
+      },
+    });
+
+    if (!expiring) {
+      throw new BadRequestException("Invalid refresh token");
+    }
+
+    const accessExpiry = DateTime.now().plus({ days: 1 }).startOf("day").toJSDate();
+    const refreshExpiry = DateTime.now().plus({ year: 1 }).startOf("day").toJSDate();
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const [_, _refresh, accessToken, refreshToken] = await this.dbWrite.prisma.$transaction([
+      this.dbWrite.prisma.accessToken.deleteMany({
+        where: { client: { id: oauthClient.id }, expiresAt: { lte: new Date() } },
```

Comment:
not sure we want to delete all the expired access token of an oAuth client, we won't be able to differentiate expired tokens still trying to be used from the ones that never existed in the first place

Reply:
Why would that be something we care about?

Reply:
I'm thinking it could be useful to have specific error messages, so that the Atoms know when it should bother trying to refresh the token or just display unauthorized error